src/tools/interop/idt/capture/platform/android/android.py

```diff
@@ -0,0 +1,220 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
+#    All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+import ipaddress
+import os
+import time
+import typing
+
+from capture.base import PlatformLogStreamer
+from capture.file_utils import create_standard_log_name
+from capture.shell_utils import BashRunner
+
+
+class Android(PlatformLogStreamer):
+    """
+    Class that supports:
+    - Running synchronous adb commands
+    - Maintaining a singleton logcat stream
+    - Maintaining a singleton screen recording
+    """
+
+    def __init__(self, artifact_dir: str) -> None:
+
+        self.artifact_dir = artifact_dir
+
+        self.device_id: str | None = None
+        self.adb_devices: typing.Dict[str, bool] | None = None
+        self._authorize_adb()
+
+        # TODO: Set a larger buffer size
+        # TODO: More efficient buffer for analysis than the file
+        self.logcat_output_path = os.path.join(
+            self.artifact_dir, create_standard_log_name(
+                'logcat', 'txt'))
+        self.logcat_proc: BashRunner | None = None
+        self.logcat_command = f'adb -s {self.device_id} logcat -T 1 >> {self.logcat_output_path}'
+
+        # TODO: BT Logs
+
+        screen_cast_name = create_standard_log_name('screencast', 'mp4')
+        self.screen_cap_output_path = os.path.join(
+            self.artifact_dir, screen_cast_name)
+        self.screen_proc: BashRunner | None = None
+        self.check_screen_command = "shell dumpsys deviceidle | grep mScreenOn"
+        self.screen_path = f'/sdcard/Movies/{screen_cast_name}'
+        self.screen_command = f'adb -s {self.device_id} shell screenrecord {self.screen_path}'
```

Comment:
In my experience, the buffer is too small. Here are my notes:

```
    Android screen recorder has a max duration of 180 seconds.
    Android screen recorder does not allow you to cap the file-size at a specific limit.

To work around these limitations, while generating files that can be attached to JIRAs, the bit-rate option can be used to attempt to limit the file-size, with the knowledge of max 180 second duration.
```

```bash
# The command is executed multiple times such that after each command reaches it's recording limit, then a new recording will be started.
adb shell screenrecord --bit-rate 1000000 --bugreport /sdcard/Download/screencap-1.mp4
adb shell screenrecord --bit-rate 1000000 --bugreport /sdcard/Download/screencap-2.mp4
adb shell screenrecord --bit-rate 1000000 --bugreport /sdcard/Download/screencap-3.mp4
adb shell screenrecord --bit-rate 1000000 --bugreport /sdcard/Download/screencap-4.mp4
```

Reply:
`--bugreport` is useful because it will show the timestamp on the screen. I use it to correlate touch event in the video to log line.

Reply:
Ack thanks this is very useful info. We'll have to handle this more elegantly moving forward.